californium-core/src/main/java/org/eclipse/californium/core/CoapClient.java

```diff
@@ -846,7 +849,15 @@ private Request ifNoneMatch(Request request) {
 	private CoapObserveRelation observeAndWait(Request request, CoapHandler handler) {
 		Endpoint outEndpoint = getEffectiveEndpoint(request);
 		CoapObserveRelation relation = new CoapObserveRelation(request, outEndpoint);
-		request.addMessageObserver(new ObserveMessageObserverImpl(handler, relation));
+		// add message observer to get the response.
+		ObserveMessageObserverImpl messageObserver = new ObserveMessageObserverImpl(handler, relation);
+		request.addMessageObserver(messageObserver);
+		// add notification listener to all notification
+		NotificationListener notificationListener = new Adapter(messageObserver, request);
+		outEndpoint.addNotificationListener(notificationListener);
+		// messageobserver should remove this listener when the request is cancelled
+		messageObserver.setNotificationListener(notificationListener);
+		relation.setNotificationListener(notificationListener);
```

Comment:
If we don't force anybody about in-memory cache. That means we can not consider that we always have a request with its `MessageObserver` in memory, right ? so you agree that we need to use NotificationListener in coapClient. (Or do I miss something ?)

About cleaning the first level cache, what about if the cancel is not done on instances which have this request  cached in memory ? (Not so important as we agree that this should not be mandatory)

About the adapter, you're right. I don't know how we can avoid this :'(... In all case I'm not sure the `CoapClient` API is well adapted when you want to handle a lot of clients, because you need to create a coapClient by [IP/port/resourceURI](https://github.com/eclipse/californium/blob/master/californium-core/src/main/java/org/eclipse/californium/core/CoapClient.java#L95). That's why we don't use it in Leshan.